File: java/L200/LeetCode0216.java
package L200;

import java.util.ArrayList;
import java.util.List;

public class LeetCode0216 {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>(k);
        dfs(res, path, k, n, 1);
        return res;
    }

    private void dfs(List<List<Integer>> res, List<Integer> path, int k, int n, int index) {
        if (k < 0 || n < 0) {
            return;
        }
        if (k == 0) {
            if (n == 0) {
                res.add(new ArrayList<>(path));
            }
            return;
        }
        for (int i = index;i < 10;++i) {
            if (n - i < 0) {
                break;
            }
            path.add(i);
            dfs(res, path, k - 1, n - i, i + 1);
            path.remove(path.size() - 1);
        }
    }
}

File: cpp/L200/LeetCode0216.cc
#include <vector>
using std::vector;

class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> res;
        vector<int> path;
        int N = 1 << 9;
        for (int x = 1;x < N;++x) {
            vector<int> path;
            int sum{0};
            for (int i = 0;i < 9;++i) {
                if (((x >> i) & 1) == 1) {
                    path.push_back(i + 1);
                    sum += i + 1;
                }
            }
            if (sum == n && path.size() == k) {
                res.push_back(path);
            }
        }
        return res;
    }
};

int main() {
    Solution s;
    s.combinationSum3(3, 7);
}